src/ComponentProvider.ts

```diff
@@ -0,0 +1,199 @@
+
+  constructor(private document: AsyncAPIDocument) {
+    this.scanChannels();
+    this.scanComponents();
+  }
+
+  /**
+   * This function is responsible for scanning the document's channels section.
+   *
+   * @returns {void}
+   */
+  private scanChannels = (): void => {
+    const channels = this.document.channels();
+    for (const channelName in channels) {
+      const path = `channels.${channelName}`;
+      const channel = this.document.channel(channelName);
+
+      if (channel.hasPublish()) {
+        const message = channel.publish().message();
+        const currentPath = `${path}.publish.message`;
+        this.scanMessage(currentPath, message);
+        this.messages.set(currentPath, message);
+      }
+
+      if (channel.hasSubscribe()) {
+        const message = channel.subscribe().message();
+        const currentPath = `${path}.subscribe.message`;
+        this.scanMessage(currentPath, message);
+        this.messages.set(currentPath, message);
+      }
+
+      if (channel.hasParameters()) {
+        const currentPath = `${path}.parameters`;
+        for (const [name, channelParameter] of Object.entries(channel.parameters())) {
+          this.parameters.set(`${currentPath}.${name}`, channelParameter);
+        }
+      }
+    }
+  }
+  /**
+   * This function is responsible for scanning any schema that is passed to it.
+   *
+   * @param {string} path - path of the schema that needs to be scanned.
+   * @param {Schema} schema - the actual schema object.
+   *
+   * @returns {void}
+   */
+  private scanSchema = (path: string, schema: Schema): void => {
+    this.schemas.set(path, schema);
+    const schemaProperties = schema.properties();
+    for (const [propertyName, propertySchema] of Object.entries(schemaProperties)) {
+      this.schemas.set(`${path}.properties.${propertyName}`, propertySchema);
+    }
+  }
+
+  /**
+   * This function will determine if a component is in its respective traits or not.
+   *
+   * @param {string} type - A string to pass the name of the child component. it should
+   * be the field name that is passed in childComponent field.
+   * @param {any} childComponent - this is the child component that needs to be checked.
+   * @param {any} parent - this is the childComponent's parent. it should contain traits filed.
+   * @returns {number} It will return 0 if childComponent is in traits and is completely taken from traits by parser.
+   * 1 if some filds were taken from traits.
+   * 2 if it was not found in traits.
+   *
+   * @example
+   *
+   *     isInTraits('payload', payload, message)
+   */
+  private isInTraits = (type: string, childComponent: any, parent: any): number => {
+    const traits = parent.extension('x-parser-original-traits');
+    if (Array.isArray(traits)) {
+      for (const trait of traits) {
+        for (const key in trait) {
+          if (key === type) {
+            if (compareComponents(childComponent.json(), trait[String(key)])) {
+              return 0;
```

Comment:
@KhudaDad414 I'll suggest to use enums here instead of numbers. Referencing and remembering numbers elsewhere is not very scalable approach.